Feat: Add auto formatting bot

georgereuben opened the pull request:
### Description

<!--
If this is your first contribution to Lucene, please make sure you have reviewed the contribution guide.
https://github.com/apache/lucene/blob/main/CONTRIBUTING.md
-->

Added a workflow bot that automatically formats a PR's file changes and raises a separate PR against that PR

### How it works :
When someone comments ```/format-fix apply``` on a PR, the bot will:
- Check your code using ./gradlew check -x test to find any formatting/style issues
- Fix what it can by running ./gradlew tidy (which includes Google Java Format, EditorConfig fixes, and other style tools)
- Verify everything looks good by running the checks again
- Create a separate PR with just the formatting fixes, so you can review and merge them without mixing formatting changes with your actual code changes

### Supported files:
Java, Gradle, Groovy, Markdown, properties, XML, Python, shell scripts, and batch files.
Just comment "/format-fix apply" on any PR to trigger it

Closes: [Issue #14835](https://github.com/apache/lucene/issues/14835) 

Test Run : [Test PR](https://github.com/georgereuben/lucene/pull/2)

georgereuben committed:
add auto format workflow

georgereuben committed:
add reactions

dweiss commented:
This looks very interesting... and way beyond my knowledge of gh actions. What are the security implications for having these permissions in the workflow? Can somebody craft a PR that would somehow alter Lucene's main repo?

georgereuben commented:
> This looks very interesting... and way beyond my knowledge of gh actions. What are the security implications for having these permissions in the workflow? Can somebody craft a PR that would somehow alter Lucene's main repo?

There is a check before the workflow runs so that only users with admin or write repository permissions can trigger the bot. I have tested it in the test PR and verified that the workflow rejects the trigger command from an external actor

<img width="1512" alt="Screenshot 2025-07-09 at 5 23 25 PM" src="https://github.com/user-attachments/assets/ace92ed0-192b-4ff4-89ed-8b710f6f8f51" />

rmuir commented on .github/workflows/auto-format.yml:
```diff
@@ -0,0 +1,359 @@
+
+    steps:
+      - name: Check permissions
+        uses: actions/github-script@v7
```
can we avoid the use of all the javascript and just use `gh` cli instead?

rmuir replied:
if we need js, then we need it, but such large scripts should be factored into separate .js files e.g. in a scripts/ directory.

inline is not good for managing large amounts of logic.

rmuir commented on .github/workflows/auto-format.yml:
```diff
@@ -0,0 +1,359 @@
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              body: comment
+            });
```
for example, for this entire set of javascript, i'd just use `gh pr comment`

georgereuben committed:
replace inline scripts with gh cli

georgereuben committed:
add repo flag to gh pr commands

dweiss commented:
> Should I go ahead and implement a check where it will allow the PR author along with folks with admin and write access to trigger the format bot?

Yes, I think this makes sense. It's a PR against their own repository (against their PR), so I think it's fine (?).

georgereuben committed:
fix: modify reactions to use double-quotes

georgereuben committed:
fix: add input field for reactions

georgereuben committed:
allow PR author to trigger the workflow

georgereuben commented:
> Right - thank you for working on this. It seems way more complex than I thought. Also, it entails some worrying uncertainty with respect to what can go wrong. Here's an example - I received an email today with this:
> but I can't locate this "comment" anywhere on this issue. Yet it did send out an e-mail, somehow. It's kind of scary to me.

Yes sorry, I was testing the gh cli comment command locally and accidentally added this PR instead of my forked repo's PR as the target. So this test comment was posted from my id on that PR. I deleted it later, my apologies.

dweiss commented on .github/workflows/auto-format.yml:
```diff
@@ -0,0 +1,268 @@
+    runs-on: ubuntu-latest
+    timeout-minutes: 30
+
+    permissions:
+      contents: write
+      pull-requests: write
+      issues: write
```
Why is the contents permission needed? Just wondering if we could omit it.

georgereuben replied:
It will be needed for pushing formatting commits

dweiss commented on .github/workflows/auto-format.yml:
```diff
@@ -0,0 +1,268 @@
+          token: ${{ secrets.GITHUB_TOKEN }}
+          fetch-depth: 0
+
+      - name: Set up JDK 21
+        uses: actions/setup-java@v4
+        with:
+          java-version: '21'
+          distribution: 'temurin'
```
this should be java 24 for the main branch. Ideally, you'd reuse Lucene's own prepare-for-build action, much like other workflows do.

georgereuben replied:
Got it thanks, fixing this

georgereuben committed:
update get-changed-files  to use trusted action

georgereuben commented:
Hi @dweiss, I now see that it might not be possible to raise a PR against a PR raised from a branch in a forked repo since the github actions workflow will not have access to that forked repo. 
The other viable solution I can see is to add a commit to the author's PR itself as this will be possible if the PR author has allowed edits by maintainers. I have been working on this but I wanted to check with you if this would be viable

georgereuben committed:
Add direct commit conditional flow for PRs from forked repos

georgereuben committed:
replace get-changed-files step with gh cli script

georgereuben commented on .github/workflows/auto-format.yml:
```diff
@@ -0,0 +1,359 @@
+
+      - name: Get changed files
+        id: changed-files
+        uses: tj-actions/changed-files@v46
```
I tried using ana06/get-changed-files@v2.3.0, but it doesn't support issue comment triggers. So I have replaced this action with a gh cli script that would perform the same action when triggered by a comment. Please let me know if this works.

georgereuben committed:
replace JDK setup with Lucene's prepare for build action

georgereuben commented:
Hi @dweiss @rmuir, I have updated the workflow. If the PR is raised in the same repo by a maintainer, it will raise a PR with formatting fixes, and if the PR is raised from a forked repo, it will directly commit the formatting fixes to the author's PR and the author can check the commit changes.

I have updated the workflow with the rest of your suggestions as well, please review and let me know if any other improvements can be made, thanks.

dweiss reviewed (approved):
Thank you for your effort, @georgereuben 

It's a hell of a lot of a complex workflow... but I think we should give it a try. If it's used - fine, if it's not used after some time, we can drop it. 

I'm going to merge it and see what happens.

dweiss commented:
Uh, sorry - it does work, it just didn't show progress while working!
